src/transport/webrtc/substream.rs

```diff
@@ -198,17 +198,20 @@ impl tokio::io::AsyncWrite for Substream {
             return Poll::Ready(Err(std::io::ErrorKind::BrokenPipe.into()));
         }
 
-        // TODO: try to coalesce multiple calls to `poll_write()` into single `Event::Message`
-
-        let num_bytes = std::cmp::min(MAX_FRAME_SIZE, buf.len());
-        let future = self.tx.reserve();
-        futures::pin_mut!(future);
-
-        let permit = match futures::ready!(future.poll(cx)) {
-            Err(_) => return Poll::Ready(Err(std::io::ErrorKind::BrokenPipe.into())),
+        let permit = match self.tx.try_reserve() {
             Ok(permit) => permit,
+            Err(err) =>
+                return match err {
+                    TrySendError::Full(_) => {
+                        cx.waker().wake_by_ref();
+                        Poll::Pending
```

Comment:
This will effectively enter into a busy loop until there's capacity in the buffer:
- we call `poll_write` and the channel is full
- we wake up the waker if there's no capacity
- the runtime calls `poll_write` again, but without any context of the `tx` being full or not
- we enter the same code path until `tx` has available capacity, wake the waker and return Pending